src/metrics/handle.rs

```diff
@@ -0,0 +1,302 @@
+    }
+}
+
+#[derive(Debug)]
+pub enum Handle {
+    Gauge(Gauge),
+    Counter(Counter),
+    Histogram(Histogram),
+}
+
+impl Handle {
+    pub(crate) fn counter() -> Self {
+        Handle::Counter(Counter::new())
+    }
+
+    pub(crate) fn increment_counter(&mut self, value: u64) {
+        match self {
+            Handle::Counter(counter) => counter.record(value),
+            _ => unreachable!(),
+        }
+    }
+
+    pub(crate) fn gauge() -> Self {
+        Handle::Gauge(Gauge::new())
+    }
+
+    pub(crate) fn update_gauge(&mut self, value: GaugeValue) {
+        match self {
+            Handle::Gauge(gauge) => gauge.record(value),
+            _ => unreachable!(),
+        }
+    }
+
+    pub(crate) fn histogram() -> Self {
+        Handle::Histogram(Histogram::new())
+    }
+
+    pub(crate) fn record_histogram(&mut self, value: f64) {
+        match self {
+            Handle::Histogram(h) => h.record(value),
+            _ => unreachable!(),
+        };
+    }
+}
+
+#[derive(Debug)]
+pub struct Histogram {
+    buckets: Box<[(f64, AtomicU32); 22]>,
+    count: AtomicU32,
+    sum: AtomicF64,
+}
+
+impl Histogram {
+    pub(crate) fn new() -> Self {
+        // Box to avoid having this large array inline to the structure, blowing
+        // out cache coherence.
+        //
+        // The sequence here is based on powers of two. Other sequences are more
+        // suitable for different distributions but since our present use case
+        // is mostly non-negative and measures smallish latencies we cluster
+        // around but never quite get to zero with an increasingly coarse
+        // long-tail.
+        let buckets = Box::new([
+            (f64::NEG_INFINITY, AtomicU32::new(0)),
+            (0.015625, AtomicU32::new(0)),
+            (0.03125, AtomicU32::new(0)),
+            (0.0625, AtomicU32::new(0)),
+            (0.125, AtomicU32::new(0)),
+            (0.25, AtomicU32::new(0)),
+            (0.5, AtomicU32::new(0)),
+            (0.0, AtomicU32::new(0)),
+            (1.0, AtomicU32::new(0)),
+            (2.0, AtomicU32::new(0)),
+            (4.0, AtomicU32::new(0)),
+            (8.0, AtomicU32::new(0)),
+            (16.0, AtomicU32::new(0)),
+            (32.0, AtomicU32::new(0)),
+            (32.0, AtomicU32::new(0)),
+            (128.0, AtomicU32::new(0)),
+            (256.0, AtomicU32::new(0)),
+            (512.0, AtomicU32::new(0)),
+            (1024.0, AtomicU32::new(0)),
+            (2048.0, AtomicU32::new(0)),
+            (4096.0, AtomicU32::new(0)),
+            (f64::INFINITY, AtomicU32::new(0)),
+        ]);
+        Self {
+            buckets,
+            count: AtomicU32::new(0),
+            sum: AtomicF64::new(0.0),
+        }
+    }
+
+    pub(crate) fn record(&mut self, value: f64) {
+        let mut prev_bound = f64::NEG_INFINITY;
+        for (bound, bucket) in self.buckets.iter_mut() {
+            if value > prev_bound && value <= *bound {
```

Comment:
You should be able to get rid of one of these comparisons. I think you were referring to this idea too here: https://github.com/timberio/vector/pull/7014/files#r608864011

Reply:
Hmm, well. After poking at this for a while I couldn't come up with any other implementation that makes a measurable difference in runtime, since I want to avoid assuming foreknowledge about bucket spacing. The logic strikes me as reasonable explicable so I'll let it alone.

Reply:
Yes, it's unlikely to make a visible difference except in microbenchmarks.